gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,431 @@
+            {
+                return false;
+            }
+        }
+        // Also skip frozen slots or slots with a newer proof than me.
+        match self.chunk_map.get(&slot) {
+            Some(SlotStatus::Frozen) => {
+                return false;
+            }
+            Some(SlotStatus::UnfinishedProof(slot_map)) => {
+                if let Some(proof_chunkmap) = slot_map.get(&data.from) {
+                    if proof_chunkmap.wallclock > data.wallclock {
+                        return false;
+                    }
+                }
+            }
+            None => {}
+        }
+        true
+    }
+
+    fn new_proof_chunk_map(num_chunks: u8, wallclock: u64) -> ProofChunkMap {
+        ProofChunkMap {
+            num_chunks,
+            missing_chunks: num_chunks,
+            chunks: HashMap::new(),
+            wallclock,
+        }
+    }
+
+    fn mark_slot_proof_received(&mut self, slot: u64) {
+        self.chunk_map.insert(slot, SlotStatus::Frozen);
+        for (_, current_slots_set) in self.validator_pending_proof_map.iter_mut() {
+            current_slots_set.remove(&slot);
+        }
+    }
+
+    fn insert_chunk(&mut self, data: DuplicateShred) -> Result<Option<Vec<DuplicateShred>>, Error> {
+        if let SlotStatus::UnfinishedProof(slot_chunk_map) = self
+            .chunk_map
+            .entry(data.slot)
+            .or_insert_with(|| SlotStatus::UnfinishedProof(HashMap::new()))
+        {
+            let proof_chunk_map = slot_chunk_map
+                .entry(data.from)
+                .or_insert_with(|| Self::new_proof_chunk_map(data.num_chunks, data.wallclock));
+            if proof_chunk_map.wallclock < data.wallclock {
+                proof_chunk_map.num_chunks = data.num_chunks;
+                proof_chunk_map.missing_chunks = data.num_chunks;
+                proof_chunk_map.wallclock = data.wallclock;
+                proof_chunk_map.chunks.clear();
+            }
+            let num_chunks = data.num_chunks;
+            let chunk_index = data.chunk_index;
+            let slot = data.slot;
+            let from = data.from;
+            if num_chunks == proof_chunk_map.num_chunks
+                && chunk_index < num_chunks
+                && !proof_chunk_map.chunks.contains_key(&chunk_index)
+            {
+                proof_chunk_map.missing_chunks = proof_chunk_map.missing_chunks.saturating_sub(1);
+                proof_chunk_map.chunks.insert(chunk_index, data);
+                if proof_chunk_map.missing_chunks == 0 {
+                    let mut result: Vec<DuplicateShred> = Vec::new();
+                    for i in 0..num_chunks {
+                        result.push(proof_chunk_map.chunks.remove(&i).unwrap())
+                    }
+                    return Ok(Some(result));
+                }
+            }
+            self.validator_pending_proof_map
+                .entry(from)
+                .or_insert_with(HashSet::new)
```

Comment:
`.or_default()` instead of `.or_insert_with(HashSet::new)`.

Reply:
done